core/core.go

```diff
@@ -1,43 +1,104 @@
+// or can be registered in pandora plugin system (look at core/plugin pkg), for creating engine from abstract config.
 package core
 
 import (
 	"context"
+	"time"
 )
 
-// TODO(skipor): doc
+// TODO(skipor): pass some dependencies from engine to core extpoints.
+// Extra argument to Run, or through context values, maybe.
 
+// Ammo is data required for one shot. Usually it contains something that differs
+// from one shot to another.
+// Something like requested recourse indetificator, query params, meta information
+// helpful for future shooting analysis.
 type Ammo interface{}
 
 //go:generate mockery -name=Provider -case=underscore -outpkg=coremock
 
+// Provider is routine that generates ammo for instance shoots.
+// A Provider must goroutine safe.
 type Provider interface {
-	Start(context.Context) error
-	// Acquire acquires ammo for shoot.
+	// Run starts provider routine. Blocks until ammo finish, error or context cancel.
+	// Run must be called once before any Acquire or Release calls.
+	// In case of context cancel, return nil (recommended), ctx.Err(), or error caused ctx.Err()
+	// in terms of github.com/pkg/errors.Cause.
+	Run(context.Context) error
+	// Acquire acquires ammo for shoot. Should be lightweight, so instance can shoot as
+	// soon as possible. That means ammo format parsing done in provider background routine,
+	// and acquire just takes ammo from ready pool.
 	// Ok false means that shooting should be stopped: ammo finished or shooting is canceled.
+	// Acquire may be called before start, but may block until start is called.
 	Acquire() (a Ammo, ok bool)
-	// Release notifies that ammo usage is finished.
+	// Release notifies that ammo usage is finished, and it can be reused.
+	// Instance should not retain references to released ammo.
 	Release(Ammo)
 }
 
+// Sample is data containing shoot report. Return code, timings, shoot meta information.
 type Sample interface{}
 
 //go:generate mockery -name=Aggregator -case=underscore -outpkg=coremock
 
+// Aggregator is routine that aggregates samples from all instances.
+// Usually aggregator is shooting result reporter, that writes released samples
+// to file in machine readable format for future analysis.
+// An Aggregator must goroutine safe.
 type Aggregator interface {
-	Start(context.Context) error
-	Release(Sample)
+	// Run starts aggregator routine. Blocks until error or context cancel.
+	// In case of context cancel, return nil, ctx.Err(), or error caused ctx.Err()
+	// in terms of github.com/pkg/errors.Cause.
+	Run(context.Context) error
+	// Report reports sample to aggregator. Should be lightweight, so instance can shoot as soon as possible.
+	// That means, that sample encode and reporting IO done in aggregator provider routine.
+	// Reported sample can be reused for efficiency.
+	// Report may be called before start, but may block until start is called.
+	Report(Sample)
```

Comment:
Even channels with large buffer can't guarantee that - channel buffer can be filled up, if Aggreggator handle reported samples too slow. Sample blocking queue  with dynamically growing  buffer can in case of infinity memory size. But we have not infinity memory usually:)
We can throw samples away, in case buffer fill up, but I have no arguments for why this is better.